ktesting: abort entire test suite on SIGINT + /dev/tty

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

When aborting an integration test with CTRL-C while it runs, the current test fails and etcd exits. But additional tests were still being started and the failed slowly because they couldn't connect to etcd.

It's better to fail additional tests in ktesting.Init when the test run has already been interrupted.

While at it, also make it a bit more obvious that testing was interrupted by logging it and update one comment about this.

Example:

    $ go test -v ./test/integration/quota
    ...
    I1106 11:42:48.857162  147325 etcd.go:416] "Not using watch cache" resource="events.events.k8s.io"
    I1106 11:42:48.857204  147325 handler.go:286] Adding GroupVersion events.k8s.io v1 to ResourceManager
    W1106 11:42:48.857209  147325 genericapiserver.go:765] Skipping API events.k8s.io/v1beta1 because it has no resources.
    ^C

    INFO: canceling context: received interrupt signal

    {"level":"warn","ts":"2024-11-06T11:42:48.984676+0100","caller":"embed/serve.go:160","msg":"stopping insecure grpc server due to error","error":"accept tcp 127.0.0.1:44177: use of closed network connection"}
    ...
    I1106 11:42:50.042430  147325 handler.go:142] kube-apiserver: GET "/apis/rbac.authorization.k8s.io/v1/clusterroles" satisfied by gorestful with webservice /apis/rbac.authorization.k8s.io/v1
        test_server.go:241: timed out waiting for the condition
    --- FAIL: TestQuota (11.45s)
    === RUN   TestQuotaLimitedResourceDenial
        quota_test.go:292: testing has been interrupted: received interrupt signal
    --- FAIL: TestQuotaLimitedResourceDenial (0.00s)
    === RUN   TestQuotaLimitService
        quota_test.go:418: testing has been interrupted: received interrupt signal
    --- FAIL: TestQuotaLimitService (0.00s)
    FAIL

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
ktesting: abort entire test suite on SIGINT

When aborting an integration test with CTRL-C while it runs,
the current test fails and etcd exits. But additional tests were still being
started and the failed slowly because they couldn't connect to etcd.

It's better to fail additional tests in ktesting.Init when the test run has
already been interrupted.

While at it, also make it a bit more obvious that testing was interrupted by
logging it and update one comment about this and clean up the naming of
contexts in the code.

Example:

    $ go test -v ./test/integration/quota
    ...
    I1106 11:42:48.857162  147325 etcd.go:416] "Not using watch cache" resource="events.events.k8s.io"
    I1106 11:42:48.857204  147325 handler.go:286] Adding GroupVersion events.k8s.io v1 to ResourceManager
    W1106 11:42:48.857209  147325 genericapiserver.go:765] Skipping API events.k8s.io/v1beta1 because it has no resources.
    ^C

    INFO: canceling test context: received interrupt signal

    {"level":"warn","ts":"2024-11-06T11:42:48.984676+0100","caller":"embed/serve.go:160","msg":"stopping insecure grpc server due to error","error":"accept tcp 127.0.0.1:44177: use of closed network connection"}
    ...
    I1106 11:42:50.042430  147325 handler.go:142] kube-apiserver: GET "/apis/rbac.authorization.k8s.io/v1/clusterroles" satisfied by gorestful with webservice /apis/rbac.authorization.k8s.io/v1
        test_server.go:241: timed out waiting for the condition
    --- FAIL: TestQuota (11.45s)
    === RUN   TestQuotaLimitedResourceDenial
        quota_test.go:292: testing has been interrupted: received interrupt signal
    --- FAIL: TestQuotaLimitedResourceDenial (0.00s)
    === RUN   TestQuotaLimitService
        quota_test.go:418: testing has been interrupted: received interrupt signal
    --- FAIL: TestQuotaLimitService (0.00s)
    FAIL

pohly commented:
/cc @googs1025 

Are you perhaps interested in ktesting? This is a small tweak, more might come based on feedback for https://kcsna2024.sched.com/event/1nSjo/unified-framework-for-unit-integration-and-e2e-testing

googs1025 commented:
> /cc @googs1025
> 
> Are you perhaps interested in ktesting? This is a small tweak, more might come based on feedback for https://kcsna2024.sched.com/event/1nSjo/unified-framework-for-unit-integration-and-e2e-testing

Thanks for the invitation, I will follow this topic 😄

pohly committed:
ktesting: print info and progress to /dev/tty

The "received interrupt signal" is useful also when running with "go test"
without -v because it shows that the shutdown has started.

But more important is that a progress report gets shown because that feature is
useful in particular when "go test" produces no output while it runs.

pohly commented:
/assign @googs1025 
/priority backlog

googs1025 commented:
/retest 
to confirm that the CI failed is due to an no related flake, not this change

aojea commented:
can we postpone this for next release to avoid adding functionality to test that may have a large blast radius when we are close to release?

pohly commented:
Yes, let's hold it for 1.33.

/hold

googs1025 commented:
/test pull-kubernetes-e2e-gce

maybe we can move to the next step ! 
/lgtm  from my side

googs1025 commented:
/remove-lifecycle stale

pohly commented:
/reopen

pohly committed:
ktesting: properly clean up

When cleaning up the progress channel properly (stop signal delivery, closing
the channel), the loop dumping progress reports no longer needs to check for
the separate shutdown context. Instead, it can distinguish between "signal
received" and "channel closed".

The signal context was getting cleanup by canceling it, but a channel is better
because it avoids the slightly misleading "received interrupt signal"
cancellation when the test was only shutting down.

nojnhuh commented on test/utils/ktesting/signals.go:
```diff
@@ -78,21 +82,49 @@ func (p *progressReporter) init(tb TB) context.Context {
+		// Opening /dev/tty during init avoids the redirection.
+		// May fail, depending on the OS, in which case
+		// os.Stderr is used.
+		if console, err := os.OpenFile("/dev/tty", os.O_RDWR|os.O_APPEND, 0); err == nil {
```
Should we call `console.Close()` when the test is over?

nojnhuh commented on test/utils/ktesting/stepcontext_test.go:
```diff
@@ -93,11 +99,21 @@ func TestProgressReport(t *testing.T) {
+INFO: canceling test context: received interrupt signal
+
+`))
+	gomega.NewGomegaWithT(t).Expect(context.Cause(tCtx)).To(gomega.MatchError(gomega.Equal("received interrupt signal")), "interrupted")
```
I begrudgingly asked Copilot and it found a race condition where the test runs
```go
defaultProgressReporter.signalChannel <- os.Interrupt
```

Then the signal handler runs
```go
_, _ = fmt.Fprint(p.out, "\n\nINFO: canceling test context: received interrupt signal\n\n")
```

In the failing case, the rest of the test runs before the next line of the signal handler runs, which the test asserts has already run.
```go
interrupted(errors.New("received interrupt signal"))
```

Copilot's fix is to call `interrupted` before `fmt.Fprint` in the signal handler. That effectively blocks the test at `message := <-out.stream` until the context is canceled. Then the test passes reliably for me.

pohly commented:
/test pull-kubernetes-e2e-gce

Cluster setup.

nojnhuh reviewed (approved):
/lgtm
/remove-lifecycle rotten

> Yes, let's hold it for 1.33.
> 
> /hold

I think we should be ok to get this in now.

/hold cancel
